File: lib/machine-reliability.js
'use strict';
(function (global) {
/**
 * Which machine is costing the shop, and what it keeps doing wrong.
 *
 * Neither app has answered this. Waste is charted by failure type over time —
 * which says the shop has a warping problem — and never by MACHINE, which is
 * what says WHICH printer has it. "Replace the old one" is a decision worth
 * thousands, and until now there was nothing to make it on.
 *
 * ── AGAINST WHAT THE MACHINE PRINTED, NOT IN ISOLATION ────────────────────
 *
 * A printer that ran nine hundred hours and scrapped two kilos is doing better
 * than one that ran ninety and scrapped one — so the rate is scrap against what
 * that machine actually put out, and the raw grams are reported beside it
 * rather than instead of it. Ranking by grams alone would always name the
 * busiest machine, which is the wrong printer to sell.
 *
 * Pure: no DOM, no fs, no Electron.
 */

function num(v) {
  const n = Number(v);
  return Number.isFinite(n) ? n : 0;
}

function day(v) { return String(v == null ? '' : v).slice(0, 10); }

const FINISHED = ['completed', 'delivered'];

/**
 * @param {object} input
 *   machines  [{ id, name, color }]
 *   orders    every order; finished ones say what each machine put out
 *   waste     [{ machineId, date, weight, cost, failureType }]
 *   from, to  `YYYY-MM-DD` bounds, or '' for all time
 *   unassigned what to call scrap that names no machine
 * @param {object} deps
 *   gramsOf  (order) => grams that machine actually printed
 *   hoursOf  (order) => hours it ran
 * @returns {{rows: Array, totals: object}} worst rate first
 */
function machineReliability(input, deps) {
  const i = input || {};
  const d = deps || {};
  const gramsOf = typeof d.gramsOf === 'function'
    ? d.gramsOf
    : (o) => (o && Array.isArray(o.parts) ? o.parts : []).reduce(
        (s, p) => s + (num(p && p.printWeight) + num(p && p.supportWeight))
                      * Math.max(1, num(p && p.qty) || 1), 0);
  const hoursOf = typeof d.hoursOf === 'function'
    ? d.hoursOf
    : (o) => (o && Array.isArray(o.parts) ? o.parts : []).reduce(
        (s, p) => s + num(p && p.printTime) * Math.max(1, num(p && p.qty) || 1), 0);

  const from = day(i.from);
  const to = day(i.to);
  const inWindow = (at) => {
    const when = day(at);
    if (!when) return false;
    if (from && when < from) return false;
    if (to && when > to) return false;
    return true;
  };

  const NONE = '__none__';
  const byId = new Map();
  for (const m of (Array.isArray(i.machines) ? i.machines : [])) {
    if (!m || !m.id) continue;
    byId.set(String(m.id), {
      machineId: String(m.id), name: String(m.name || ''), color: m.color || '#888888',
      jobs: 0, grams: 0, hours: 0,
      scraps: 0, scrapGrams: 0, scrapCost: 0,
      scrapRate: null, worstFault: null, faults: {},
    });
  }
  // Scrap that names no machine is still scrap. A shop cannot act on it, but
  // hiding it makes the shop's total look better than it is.
  byId.set(NONE, {
    machineId: NONE, name: String(i.unassigned || ''), color: '#888888',
    jobs: 0, grams: 0, hours: 0,
    scraps: 0, scrapGrams: 0, scrapCost: 0,
    scrapRate: null, worstFault: null, faults: {},
  });

  for (const order of (Array.isArray(i.orders) ? i.orders : [])) {
    if (!order || order.voidedAt) continue;
    if (!FINISHED.includes(String(order.status || ''))) continue;
    if (!inWindow(order.date)) continue;
    const row = byId.get(String(order.machineId || '')) || byId.get(NONE);
    row.jobs += 1;
    row.grams += num(gramsOf(order));
    row.hours += num(hoursOf(order));
  }

  for (const entry of (Array.isArray(i.waste) ? i.waste : [])) {
    if (!entry || !inWindow(entry.date)) continue;
    const row = byId.get(String(entry.machineId || '')) || byId.get(NONE);
    row.scraps += 1;
    row.scrapGrams += num(entry.weight);
    row.scrapCost += num(entry.cost);
    const fault = String(entry.failureType || 'other');
    row.faults[fault] = num(row.faults[fault]) + num(entry.weight);
  }

  for (const row of byId.values()) {
    // Against what the machine PUT OUT, plus what it scrapped — the denominator
    // is everything it consumed, so a machine that scrapped half its filament
    // reads as 50% rather than 100%.
    const handled = row.grams + row.scrapGrams;
    row.scrapRate = handled > 0 ? row.scrapGrams / handled : null;
    // What it keeps doing wrong, which is the actionable half: "warping" sends
    // somebody to the chamber temperature, and a number does not.
    const faults = Object.entries(row.faults).sort((a, b) => b[1] - a[1]);
    row.worstFault = faults.length ? { type: faults[0][0], grams: faults[0][1] } : null;
  }

  const rows = [...byId.values()]
    .filter((r) => r.jobs > 0 || r.scraps > 0)
    // WORST RATE FIRST, not most grams — ranking by grams always names the
    // busiest machine, which is the wrong printer to sell.
    .sort((a, b) => (b.scrapRate ?? -1) - (a.scrapRate ?? -1)
                    || b.scrapGrams - a.scrapGrams);

  const grams = rows.reduce((s, r) => s + r.grams, 0);
  const scrapGrams = rows.reduce((s, r) => s + r.scrapGrams, 0);
  const handled = grams + scrapGrams;
  return {
    rows,
    totals: {
      jobs: rows.reduce((s, r) => s + r.jobs, 0),
      grams, scrapGrams,
      scraps: rows.reduce((s, r) => s + r.scraps, 0),
      scrapCost: rows.reduce((s, r) => s + r.scrapCost, 0),
      scrapRate: handled > 0 ? scrapGrams / handled : null,
      /// The machine to look at — worst rate, and only once it has printed
      /// enough for a rate to mean anything. One scrapped print on a machine
      /// that has run twice is not evidence of anything.
      worst: rows.find((r) => r.scrapRate != null && r.scraps > 0 && r.jobs >= 2) || null,
    },
  };
}

const api = { machineReliability, FINISHED };
if (typeof module !== 'undefined' && module.exports) module.exports = api;
global.KhaytMachineReliability = api;
})(typeof globalThis !== 'undefined' ? globalThis : this);

File: mac/KhaytCore/Sources/KhaytCore/MachineReliability.swift
import Foundation

/// Which machine is costing the shop, and what it keeps doing wrong.
///
/// Neither app has answered this. Waste is charted by failure type over time —
/// which says the shop has a warping problem — and never by MACHINE, which is
/// what says WHICH printer has it. "Replace the old one" is a decision worth
/// thousands, and until now there was nothing to make it on.
///
/// ── AGAINST WHAT THE MACHINE PRINTED, NOT IN ISOLATION ────────────────────
///
/// A printer that ran nine hundred hours and scrapped two kilos is doing better
/// than one that ran ninety and scrapped one — so the rate is scrap against
/// what that machine actually put out, and the raw grams are reported beside it
/// rather than instead of it. Ranking by grams alone would always name the
/// busiest machine, which is the wrong printer to sell.
public enum MachineReliability {

    public static let finished = ["completed", "delivered"]

    /// The key scrap that names no machine is gathered under.
    public static let noMachine = "__none__"

    static func num(_ value: JSONValue?) -> Double {
        let n = JSSemantics.number(value)
        return n.isFinite ? n : 0
    }

    static func day(_ value: JSONValue?) -> String {
        String(decoding: Array(JSSemantics.text(value).utf16.prefix(10)), as: UTF16.self)
    }

    /// `a < b` on two JavaScript strings — UTF-16 code units.
    static func less(_ a: String, _ b: String) -> Bool {
        var l = a.utf16.makeIterator(), r = b.utf16.makeIterator()
        while true {
            switch (l.next(), r.next()) {
            case (nil, nil): return false
            case (nil, _): return true
            case (_, nil): return false
            case (let x?, let y?): if x != y { return x < y }
            }
        }
    }

    /// The default `gramsOf`: every part's printed and support weight, times
    /// its quantity — `Math.max(1, qty || 1)`, so a part with no quantity
    /// counts once rather than none.
    public static func grams(of order: JSONValue) -> Double {
        parts(of: order).reduce(0) { sum, p in
            guard case .object(let part) = p else { return sum }
            return sum + (num(part["printWeight"]) + num(part["supportWeight"]))
                * Swift.max(1, num(part["qty"]) == 0 ? 1 : num(part["qty"]))
        }
    }

    public static func hours(of order: JSONValue) -> Double {
        parts(of: order).reduce(0) { sum, p in
            guard case .object(let part) = p else { return sum }
            return sum + num(part["printTime"])
                * Swift.max(1, num(part["qty"]) == 0 ? 1 : num(part["qty"]))
        }
    }

    private static func parts(of order: JSONValue) -> [JSONValue] {
        guard case .object(let o) = order, case .array(let list)? = o["parts"] else { return [] }
        return list
    }

    public struct Fault: Sendable, Equatable {
        public let type: String
        public let grams: Double
    }

    public struct Row: Sendable, Equatable, Identifiable {
        public let machineId: String
        public let name: String
        public let color: String
        public let jobs: Int
        public let grams: Double
        public let hours: Double
        public let scraps: Int
        public let scrapGrams: Double
        public let scrapCost: Double
        /// Nil for a machine that has handled nothing — which is not a machine
        /// with a perfect record.
        public let scrapRate: Double?
        /// What it keeps doing wrong, which is the actionable half: "warping"
        /// sends somebody to the chamber temperature, and a number does not.
        public let worstFault: Fault?
        public var id: String { machineId }
    }

    public struct Totals: Sendable, Equatable {
        public let jobs: Int
        public let grams: Double
        public let scrapGrams: Double
        public let scraps: Int
        public let scrapCost: Double
        public let scrapRate: Double?
        /// The machine to look at — worst rate, and only once it has printed
        /// enough for a rate to mean anything. One scrapped print on a machine
        /// that has run twice is not evidence of anything.
        public let worst: Row?
    }

    public struct Report: Sendable, Equatable {
        public let rows: [Row]
        public let totals: Totals
    }

    public static func report(machines: [JSONValue], orders: [JSONValue],
                              waste: [JSONValue], from: String = "", to: String = "",
                              unassigned: String = "",
                              gramsOf: ((JSONValue) -> Double)? = nil,
                              hoursOf: ((JSONValue) -> Double)? = nil) -> Report {
        let gramsOf = gramsOf ?? grams(of:)
        let hoursOf = hoursOf ?? hours(of:)
        let from = day(.string(from)), to = day(.string(to))
        func inWindow(_ at: JSONValue?) -> Bool {
            let when = day(at)
            guard !when.isEmpty else { return false }
            if !from.isEmpty && less(when, from) { return false }
            if !to.isEmpty && less(to, when) { return false }
            return true
        }

        // Insertion-ordered, because the sort is stable in the original.
        var order: [String] = []
        var rows: [String: (name: String, color: String, jobs: Int, grams: Double,
                            hours: Double, scraps: Int, scrapGrams: Double,
                            scrapCost: Double)] = [:]
        var faults: [String: [(type: String, grams: Double)]] = [:]

        for machine in machines {
            guard case .object(let m) = machine, let id = m["id"],
                  JSSemantics.truthy(id) else { continue }
            let key = JSSemantics.text(id)
            guard rows[key] == nil else { continue }
            rows[key] = (JSSemantics.truthy(m["name"]) ? JSSemantics.text(m["name"]) : "",
                         JSSemantics.truthy(m["color"]) ? JSSemantics.text(m["color"]) : "#888888",
                         0, 0, 0, 0, 0, 0)
            faults[key] = []
            order.append(key)
        }
        // Scrap that names no machine is still scrap. A shop cannot act on it,
        // but hiding it makes the shop's total look better than it is.
        if rows[noMachine] == nil {
            rows[noMachine] = (unassigned, "#888888", 0, 0, 0, 0, 0, 0)
            faults[noMachine] = []
            order.append(noMachine)
        }

        for job in orders {
            guard JSSemantics.truthy(job), case .object(let o) = job,
                  !JSSemantics.truthy(o["voidedAt"]),
                  finished.contains(JSSemantics.text(o["status"])),
                  inWindow(o["date"]) else { continue }
            let named = JSSemantics.truthy(o["machineId"]) ? JSSemantics.text(o["machineId"]) : ""
            let key = rows[named] != nil ? named : noMachine
            rows[key]!.jobs += 1
            rows[key]!.grams += num(.number(gramsOf(job)))
            rows[key]!.hours += num(.number(hoursOf(job)))
        }

        for entry in waste {
            guard JSSemantics.truthy(entry), case .object(let w) = entry,
                  inWindow(w["date"]) else { continue }
            let named = JSSemantics.truthy(w["machineId"]) ? JSSemantics.text(w["machineId"]) : ""
            let key = rows[named] != nil ? named : noMachine
            rows[key]!.scraps += 1
            rows[key]!.scrapGrams += num(w["weight"])
            rows[key]!.scrapCost += num(w["cost"])
            let fault = JSSemantics.truthy(w["failureType"])
                ? JSSemantics.text(w["failureType"]) : "other"
            if let at = faults[key]!.firstIndex(where: { $0.type == fault }) {
                faults[key]![at].grams += num(w["weight"])
            } else {
                faults[key]!.append((fault, num(w["weight"])))
            }
        }

        let built: [Row] = order.compactMap { key in
            guard let r = rows[key] else { return nil }
            // The denominator is everything the machine CONSUMED — what it put
            // out plus what it scrapped — so a machine that scrapped half its
            // filament reads as 50% rather than 100%.
            let handled = r.grams + r.scrapGrams
            let ranked = (faults[key] ?? []).enumerated().sorted { lhs, rhs in
                if lhs.element.grams != rhs.element.grams {
                    return lhs.element.grams > rhs.element.grams
                }
                return lhs.offset < rhs.offset   // Swift's sort is not stable
            }
            return Row(machineId: key, name: r.name, color: r.color, jobs: r.jobs,
                       grams: r.grams, hours: r.hours, scraps: r.scraps,
                       scrapGrams: r.scrapGrams, scrapCost: r.scrapCost,
                       scrapRate: handled > 0 ? r.scrapGrams / handled : nil,
                       worstFault: ranked.first.map { Fault(type: $0.element.type,
                                                            grams: $0.element.grams) })
        }

        let kept = built.enumerated().filter { $0.element.jobs > 0 || $0.element.scraps > 0 }
        // WORST RATE FIRST, not most grams — ranking by grams always names the
        // busiest machine, which is the wrong printer to sell.
        let ranked = kept.sorted { lhs, rhs in
            let a = lhs.element.scrapRate ?? -1, b = rhs.element.scrapRate ?? -1
            if a != b { return a > b }
            if lhs.element.scrapGrams != rhs.element.scrapGrams {
                return lhs.element.scrapGrams > rhs.element.scrapGrams
            }
            return lhs.offset < rhs.offset
        }.map(\.element)

        let grams = ranked.reduce(0) { $0 + $1.grams }
        let scrapGrams = ranked.reduce(0) { $0 + $1.scrapGrams }
        let handled = grams + scrapGrams
        return Report(rows: ranked, totals: Totals(
            jobs: ranked.reduce(0) { $0 + $1.jobs },
            grams: grams, scrapGrams: scrapGrams,
            scraps: ranked.reduce(0) { $0 + $1.scraps },
            scrapCost: ranked.reduce(0) { $0 + $1.scrapCost },
            scrapRate: handled > 0 ? scrapGrams / handled : nil,
            worst: ranked.first { $0.scrapRate != nil && $0.scraps > 0 && $0.jobs >= 2 }))
    }
}
